cg/smart_contracts/ClearGDPR.sol

```diff
@@ -1,135 +1,152 @@
+        return subjects[_subjectId].statePerProcessor[_processor];
+    }
 
-  function getProcessors() public view returns(address[], uint256){
-    return (processors, processors.length);
-  }
+    function getProcessors() public view returns(address[], uint256){
+        return (processors, processors.length);
+    }
 
-  function setSubjectDataState(bytes32 _subjectIdHash, address _processorIdHash, State _state) public returns(bool){
-    subjects[_subjectIdHash].statePerProcessor[_processorIdHash] = _state;
-    return true;
-  }
+    function setSubjectDataState(bytes32 _subjectId, address _processor, State _state) public returns(bool){
+        subjects[_subjectId].statePerProcessor[_processor] = _state;
+        return true;
+    }
 
-  function setProcessors(address[] _newProcessors) public returns(bool){ //BUG IN HERE, CAN'T USE BYTES32
-    processors = new address[](_newProcessors.length);
-    processors = _newProcessors;
-    return true;
-  }
-  
-  function isProcessor(address _processor) public view returns(bool){ //0x1 for remix, since it's the only argument
-    for(uint256 i = 0; i < processors.length; i++){
-      if(processors[i] == _processor){
+    function setProcessors(address[] _newProcessors) public returns(bool){
+        processors = new address[](_newProcessors.length + 1);
+        processors[0] = controller;
+        for(uint256 i = 0; i < _newProcessors.length; i++){
+            require(_newProcessors[i] != controller);
+            for(uint256 j = i + 1; j < _newProcessors.length; j++){ // Check if the processors are unique
+                require(_newProcessors[i] != _newProcessors[j]);
+            }
+            processors[i + 1] = _newProcessors[i];
+        }
         return true;
-      }
+    } 
+    //processors = _newProcessors; //Interesting line
+  
+    function isProcessor(address _processor) public view returns(bool){
+        for(uint256 i = 0; i < processors.length; i++){
+            if(processors[i] == _processor){
+                return true;
+            }
+        }
+        return false;
     }
-    return false;
-  }
   
-  function areAllValidProcessors(address[] _processors) public view returns(bool){
-    for(uint256 i = 0; i < _processors.length; i++){
-      require(isProcessor(_processors[i]));
+    function areAllValidProcessors(address[] _processors) public view returns(bool){
+        for(uint256 i = 0; i < _processors.length; i++){
+            require(isProcessor(_processors[i]));
+        }
+        return true;
     }
-    return true;
-  }
   
-  function recordProcessorsUpdate(address[] _newProcessors) public onlyController returns(bool){
-    require(setProcessors(_newProcessors));
-    emit Controller_ProcessorsUpdated(_newProcessors);  
-    return true;
-  }
+    function recordProcessorsUpdate(address[] _newProcessors) public onlyController returns(bool){
+        require(setProcessors(_newProcessors));
```

Comment:
Same question as above for `require`

Reply:
This require checks if the processors were updated with success. It does not return false if it fails, the entire functions reverts and we get a revert exception in the Ethereum Virtual Machine